libsolidity/analysis/TypeChecker.cpp

```diff
@@ -327,17 +329,23 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 {
 	if (_function.markedVirtual())
 	{
-		if (_function.annotation().contract->isInterface())
+		if (_function.isFree())
+			m_errorReporter.syntaxError(4493_error, _function.location(), "Free functions cannot be virtual.");
+		else if (_function.annotation().contract->isInterface())
```

Comment:
This check and a few of the others in the vicinity actually belong in the SyntaxChecker, don't they? But can be done in a separate PR.